File: UI/travle_core/lib/src/models/destination_response.dart
import 'package:json_annotation/json_annotation.dart';

import 'destination_image_response.dart';
import 'tag_ref.dart';

part 'destination_response.g.dart';

/// A tourist destination in the single shape every read path returns — list card,
/// detail, and edit prefill (mirrors the backend `DestinationResponse`). Reference
/// fields are flattened to names (never raw ids shown) and [status] is the enum
/// name. Only [primaryThumbnail] carries bytes (a small list-card thumbnail);
/// [images] is metadata, with full bytes fetched from the image endpoint.
@JsonSerializable()
class DestinationResponse {
  DestinationResponse({
    required this.id,
    required this.name,
    required this.description,
    required this.categoryId,
    required this.cityId,
    required this.latitude,
    required this.longitude,
    required this.status,
    required this.isFeatured,
    required this.averageRating,
    required this.viewCount,
    required this.submittedByUserId,
    required this.createdAt,
    this.entranceFee,
    this.categoryName,
    this.cityName,
    this.regionName,
    this.countryName,
    this.submittedByUsername,
    this.moderatedByUserId,
    this.moderatedByUsername,
    this.moderatedAt,
    this.rejectionReason,
    this.tags = const [],
    this.images = const [],
    this.primaryThumbnail,
    this.primaryThumbnailContentType,
    this.modifiedAt,
  });

  final int id;
  final String name;
  final String description;

  final int categoryId;
  final String? categoryName;

  final int cityId;
  final String? cityName;
  final String? regionName;
  final String? countryName;

  final double latitude;
  final double longitude;

  /// Optional entrance fee (KM) paid at the destination — never part of a tour's
  /// price. Null = free/unknown; the amount is an approximate "bring around X" guide.
  final double? entranceFee;

  /// Pending / Approved / Rejected — the enum name, never a raw int.
  final String status;

  final bool isFeatured;
  final double averageRating;
  final int viewCount;

  final int submittedByUserId;
  final String? submittedByUsername;

  final int? moderatedByUserId;
  final String? moderatedByUsername;
  final DateTime? moderatedAt;
  final String? rejectionReason;

  final List<TagRef> tags;
  final List<DestinationImageResponse> images;

  /// Base64 thumbnail bytes for the list card (null when there is no image).
  final String? primaryThumbnail;
  final String? primaryThumbnailContentType;

  final DateTime createdAt;
  final DateTime? modifiedAt;

  bool get isPending => status == 'Pending';
  bool get isApproved => status == 'Approved';
  bool get isRejected => status == 'Rejected';

  factory DestinationResponse.fromJson(Map<String, dynamic> json) =>
      _$DestinationResponseFromJson(json);

  Map<String, dynamic> toJson() => _$DestinationResponseToJson(this);
}

File: Backend/Travle.Model/Responses/DestinationResponse.cs
namespace Travle.Model.Responses
{
    /// <summary>
    /// A tourist destination, in the single shape used by every read path — the paginated list, the
    /// detail view, and edit prefill. Reference fields are flattened to names (never raw ids on screen)
    /// and <see cref="Status"/> is the enum name. Only the <see cref="PrimaryThumbnail"/> (a small
    /// ~15–30 KB thumbnail for list cards) carries bytes; the <see cref="Images"/> collection is
    /// metadata only, and full image bytes come from the dedicated image endpoint (§8.2 / rule 12).
    /// </summary>
    public class DestinationResponse
    {
        public int Id { get; set; }

        public string Name { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;

        public int CategoryId { get; set; }
        public string? CategoryName { get; set; }

        public int CityId { get; set; }
        public string? CityName { get; set; }
        public string? RegionName { get; set; }
        public string? CountryName { get; set; }

        public double Latitude { get; set; }
        public double Longitude { get; set; }

        /// <summary>
        /// Optional informative entrance fee (KM) paid at the destination — never part of a tour's price.
        /// Null = free/unknown; treat the amount as an approximate "bring around X" guide.
        /// </summary>
        public decimal? EntranceFee { get; set; }

        /// <summary>Pending / Approved / Rejected — the enum name, never the raw int.</summary>
        public string Status { get; set; } = string.Empty;

        public bool IsFeatured { get; set; }
        public double AverageRating { get; set; }

        /// <summary>Number of non-removed reviews behind <see cref="AverageRating"/> (shown as "(N reviews)").</summary>
        public int ReviewCount { get; set; }

        public int ViewCount { get; set; }

        /// <summary>Whether the current user has this destination in their favorites (drives the heart state).</summary>
        public bool IsFavorite { get; set; }

        public int SubmittedByUserId { get; set; }
        public string? SubmittedByUsername { get; set; }

        public int? ModeratedByUserId { get; set; }
        public string? ModeratedByUsername { get; set; }
        public DateTime? ModeratedAt { get; set; }
        public string? RejectionReason { get; set; }

        /// <summary>Tags (id + name) — ids let the edit form preselect chips; names are shown.</summary>
        public List<TagRef> Tags { get; set; } = new List<TagRef>();

        /// <summary>Image metadata (ids/order); fetch each full image via the image endpoint.</summary>
        public List<DestinationImageResponse> Images { get; set; } = new List<DestinationImageResponse>();

        /// <summary>The primary image's thumbnail bytes for list cards (null when the destination has no image).</summary>
        public byte[]? PrimaryThumbnail { get; set; }
        public string? PrimaryThumbnailContentType { get; set; }

        public DateTime CreatedAt { get; set; }
        public DateTime? ModifiedAt { get; set; }
    }
}
